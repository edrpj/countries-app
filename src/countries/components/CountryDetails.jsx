import { getCurrencyName } from "../helpers/getCurrencyName";
import { getLanguagesOfACountry } from "../helpers/getLanguagesOfACountry";
import { internationalNumberFormat } from "../helpers/internationalNumberFormat";

export const CountryDetails = ({
  name,
  nativeName,
  population,
  region,
  subregion,
  capital,
  topLevelDomain,
  currencies,
  languages,
  textColor,
}) => {
  return (
    <div className="row ms-0">
      <div className="row h-25 m-0 p-0">
        <div className="col ps-0 ps-lg-5 pt-4">
          <h2 className={`${textColor} font-extrabold`}>{name}</h2>
        </div>
      </div>

      <div className="row h-75 m-0 p-0 pt-3">
        <div className="col-12 col-sm-6 ps-0 ps-lg-5">
          <p aria-label="paragraph" className={`${textColor} mb-2`}>
            <span className="text-base font-semibold">Native Name: </span>{" "}
            {nativeName ?? name}
          </p>
          <p aria-label="paragraph" className={`${textColor} mb-2`}>
            <span className="text-base font-semibold">Population: </span>{" "}
            {internationalNumberFormat.format(population)}
          </p>
          <p aria-label="paragraph" className={`${textColor} mb-2`}>
            <span className="text-base font-semibold">Region: </span> {region}
          </p>
          <p aria-label="paragraph" className={`${textColor} mb-2`}>
            <span className="text-base font-semibold">Sub Region: </span>{" "}
            {subregion}
          </p>
          <p aria-label="paragraph" className={`${textColor} mb-2`}>
            <span className="text-base font-semibold">Capital: </span> {capital}
          </p>
        </div>

        <div className="col-12 col-sm-6 ps-0 ps-sm-4 ps-md-2 mt-4 mt-sm-0">
          <p aria-label="paragraph" className={`${textColor} mb-2`}>
            <span className="text-base font-semibold">Top Level Domain: </span>{" "}
            {topLevelDomain}
          </p>
          <p aria-label="paragraph" className={`${textColor} mb-2`}>
            <span className="text-base font-semibold">Currencies: </span>{" "}
            {getCurrencyName(currencies)}
          </p>
          <p aria-label="paragraph" className={`${textColor} mb-2`}>
            <span className="text-base font-semibold">Languages: </span>{" "}
            {getLanguagesOfACountry(languages)}
          </p>
        </div>
      </div>
    </div>
  );
};
