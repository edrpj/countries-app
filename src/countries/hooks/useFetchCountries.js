import { useEffect, useState } from "react";
import { getCountries } from "../services/getCountries";

export const useFetchCountries = (query) => {
  const [countries, setCountries] = useState([]);
  const [isLoading, setIsLoading] = useState(true);

  const fetchCountries = async () => {
    try {
      const countriesList = await getCountries(query);
      setCountries(countriesList);
      setIsLoading(false);
    } catch (error) {
      setCountries([]);
      setIsLoading(false);
    }
  };

  useEffect(() => {
    fetchCountries();
  }, [query]);

  return {
    countries,
    isLoading,
  };
};
