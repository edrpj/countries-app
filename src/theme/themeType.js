export const themeType = {
  light: "LIGHT",
  dark: "DARK",
};
