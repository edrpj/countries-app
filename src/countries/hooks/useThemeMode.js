import { useContext } from "react";

import { ThemeContext } from "../../theme/ThemeContext";
import { themeType } from "../../theme/themeType";
import { colorType } from "../types/colorType";

const {
  bgLightMode,
  bgElementsDarkMode,
  bgContainerDarkMode,
  bgBtnLightMode,
  textDarkMode,
  textLightMode,
  textInputDarkMode,
  textInputLightMode,
} = colorType;

export const useThemeMode = () => {
  const themeContextValue = useContext(ThemeContext);
  const { theme, onChangeTheme } = themeContextValue;

  const themeMode = localStorage.getItem("mode");
  const isLight = themeMode === themeType.light;

  const bgColorContainer = isLight ? bgLightMode : bgContainerDarkMode;
  const bgColorElements = isLight ? bgLightMode : bgElementsDarkMode;
  const bgColorBtn = isLight ? bgBtnLightMode : bgElementsDarkMode;
  const textColor = isLight ? textLightMode : textDarkMode;
  const textInputColor = isLight ? textInputLightMode : textInputDarkMode;

  return {
    bgColorContainer,
    bgColorElements,
    bgColorBtn,
    textColor,
    textInputColor,
    isLight,
    onChangeTheme,
    theme,
  };
};
