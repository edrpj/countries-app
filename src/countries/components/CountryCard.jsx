import { useNavigate } from "react-router-dom";
import { internationalNumberFormat } from "../helpers/internationalNumberFormat";

import { useThemeMode } from "../hooks/useThemeMode";

export const CountryCard = ({
  id,
  name,
  capital,
  population,
  region,
  imgUrl,
}) => {
  const navigate = useNavigate();
  const { bgColorElements, textColor } = useThemeMode();

  const onNavigateToACard = () => {
    navigate(`/country/${id}`);
  };

  return (
    <div className="col">
      <div
        aria-label="card"
        className={`card shadow cursor-pointer ${bgColorElements}`}
        onClick={onNavigateToACard}
      >
        <img
          src={imgUrl}
          className="card-img-top w-100 object-cover h-img-card"
          alt={name}
        />

        <div aria-label="information" className={`card-body ${textColor}`}>
          <h5 className="card-title text-base font-extrabold mb-3">{name}</h5>
          <p className="card-text mb-0">
            <span className="text-sm font-extrabold">Population: </span>{" "}
            {internationalNumberFormat.format(population)}
          </p>
          <p className="card-text mb-0">
            <span className="text-sm font-extrabold">Region: </span> {region}
          </p>
          <p className="card-text mb-3">
            <span className="text-sm font-extrabold">Capital: </span> {capital}
          </p>
        </div>
      </div>
    </div>
  );
};
