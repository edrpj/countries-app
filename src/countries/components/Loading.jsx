import { useThemeMode } from "../hooks/useThemeMode";

export const Loading = () => {
  const { isLight, bgColorContainer } = useThemeMode();

  return (
    <div
      className={`w-100 h-100 m-0-auto d-flex justify-content-center flex-1 ${bgColorContainer}`}
    >
      <div
        aria-label="spinner"
        className={`spinner-border ${isLight ? "text-dark" : "text-warning"}`}
        style={{ width: "5rem", height: "5rem" }}
        role="status"
      >
        <span className="visually-hidden">Loading...</span>
      </div>
    </div>
  );
};
