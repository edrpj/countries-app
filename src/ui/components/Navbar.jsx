import { Link } from "react-router-dom";

import { useThemeMode } from "../../countries/hooks/useThemeMode";

export const Navbar = () => {
  const { bgColorElements, textColor, onChangeTheme, theme } = useThemeMode();

  return (
    <nav className={`navbar shadow-sm py-3 ${bgColorElements}`}>
      <div className="container-fluid px-0 w-90 m-auto">
        <Link
          aria-label="logo"
          className={`navbar-brand font-extrabold ${textColor}`}
          to="countries"
        >
          Where in the world?
        </Link>
        <div
          aria-label="changeThemeDiv"
          className="d-flex align-items-center cursor-pointer"
          onClick={() => onChangeTheme(theme)}
        >
          <i
            aria-label="iTag"
            className={`fa-regular fa-moon d-flex align-items-center ${textColor}`}
          ></i>
          <p
            aria-label="paragraph"
            className={`${textColor} font-semibold ms-2 m-0-auto`}
          >
            Dark Mode
          </p>
        </div>
      </div>
    </nav>
  );
};
