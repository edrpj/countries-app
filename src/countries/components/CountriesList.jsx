import { useFetchCountries } from "../hooks/useFetchCountries";
import { CountryCard } from "./CountryCard";
import { Loading } from "./Loading";

export const CountriesList = ({ query }) => {
  const { countries, isLoading } = useFetchCountries(query);

  if (isLoading) return <Loading />;

  return (
    <div className="row row-cols-1 row-cols-sm-2 row-cols-md-3 row-cols-xl-4 g-5">
      {countries.map((country) => {
        const { cca3, name, population, region, capital, flags } = country;

        return (
          <CountryCard
            key={cca3}
            id={cca3}
            name={name.common}
            capital={capital}
            population={population}
            region={region}
            imgUrl={flags.svg}
          />
        );
      })}
    </div>
  );
};
