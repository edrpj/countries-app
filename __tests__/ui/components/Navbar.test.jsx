import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter, Routes, Route } from "react-router-dom";
import { useThemeMode } from "../../../src/countries/hooks/useThemeMode";
import { themeType } from "../../../src/theme/themeType";
import { Navbar } from "../../../src/ui/components/Navbar";

jest.mock("../../../src/countries/hooks/useThemeMode");

describe("Testing <Navbar />", () => {
  const useThemeModeData = {
    bgColorElements: "bg-white",
    textColor: "text-very-dark-blue",
    onChangeTheme: jest.fn(),
    theme: themeType.light,
  };
  useThemeMode.mockReturnValue(useThemeModeData);

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should do match with snapshot", () => {
    const { container } = render(
      <MemoryRouter>
        <Navbar />
      </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
  });

  it("should set the css classes provided by the useThemeMode custom hook", () => {
    render(
      <MemoryRouter>
        <Navbar />
      </MemoryRouter>
    );

    const { bgColorElements, textColor } = useThemeModeData;

    const nav = screen.getByRole("navigation");
    expect(nav.className).toContain(bgColorElements);

    const link = screen.getByLabelText("logo");
    expect(link.className).toContain(textColor);

    const iTag = screen.getByLabelText("iTag");
    expect(iTag.className).toContain(textColor);

    const paragraph = screen.getByLabelText("paragraph");
    expect(paragraph.className).toContain(textColor);
  });

  it("should call the onChangeTheme function with the actual theme mode", () => {
    render(
      <MemoryRouter>
        <Navbar />
      </MemoryRouter>
    );

    const changeThemeDiv = screen.getByLabelText("changeThemeDiv");
    fireEvent.click(changeThemeDiv);

    const { onChangeTheme, theme } = useThemeModeData;
    expect(onChangeTheme).toHaveBeenCalledWith(theme);
  });

  it("should navigate to the homepage when the logo is clicked", () => {
    const countriesH1Msg = "Countries";
    render(
      <MemoryRouter initialEntries={["/"]}>
        <Routes>
          <Route path="/" element={<Navbar />} />
          <Route path="countries" element={<h1>{countriesH1Msg}</h1>} />
        </Routes>
      </MemoryRouter>
    );

    const logo = screen.getByLabelText("logo");
    fireEvent.click(logo);

    expect(screen.getByText(countriesH1Msg));
  });
});
