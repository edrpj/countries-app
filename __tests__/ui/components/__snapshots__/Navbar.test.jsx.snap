// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <Navbar /> should do match with snapshot 1`] = `
<div>
  <nav
    class="navbar shadow-sm py-3 bg-white"
  >
    <div
      class="container-fluid px-0 w-90 m-auto"
    >
      <a
        aria-label="logo"
        class="navbar-brand font-extrabold text-very-dark-blue"
        href="/countries"
      >
        Where in the world?
      </a>
      <div
        aria-label="changeThemeDiv"
        class="d-flex align-items-center cursor-pointer"
      >
        <i
          aria-label="iTag"
          class="fa-regular fa-moon d-flex align-items-center text-very-dark-blue"
        />
        <p
          aria-label="paragraph"
          class="text-very-dark-blue font-semibold ms-2 m-0-auto"
        >
          Dark Mode
        </p>
      </div>
    </div>
  </nav>
</div>
`;
